MQTT Service private preview user documentation (#2695)
SoftwareAG/c8y-docs: content/device-integration/mqtt-service-bundle/implementation.md

Reviewer: An end user is more likely to know the meaning of "Notifications 2.0" than "Messaging Service" - while we should keep saying that MQTT maps to topics in the messaging service, can we have a few name-drops of N2.0 throughout this doc? Or do we not want to 'reveal' that you can go from device -> external N2.0 subscriber ?

Author: I don't really want to lead users towards N2.0 here - especially as we don't (and don't plan to) document how to create that N2.0 subscriber.

Reviewer: So long term the only supported way to publish data and subscribe to the service is via actual MQTT, Apama (soon) or the Java library that hides away N2.0?

Author: Medium-term, yes. Long-term, we will have the "Horizon 2" protocol, whatever that turns out to be, maybe even native Pulsar. We don't want to lock people in to the current N2.0 protocol that we might ultimately want to replace.

Reviewer: Fair enough. I might be a bit peeved if my only option for a custom service was Java locked into a specific SDK, but that's just me :-)

---
weight: 20
title: MQTT protocol implementation
layout: redirect
---

This section lists the implementation details for the MQTT Service.
The MQTT Service implementation supports clients connecting using MQTT versions 3.1.1 and 5.0, although not all MQTT 5.0 protocol features are supported.

### Connecting via MQTT {#connecting-via-mqtt}

MQTT Service is supported via TCP. Use your tenant domain as the URL.

Available ports:

| &nbsp; | TCP  |
|:-------|:-----|
| TLS    | 9883 |
| no TLS | 2883 |

Port 9883 (TLS) is enabled by default and should be used for secure communication.
Both one-way (server certificate only) and two-way (both client and server certificates) TLS are supported.
When client certificates are not used, the server authenticates the client using standard username and password credentials.

> Port 2883 (no TLS) is disabled by default due to security risks, as traffic is unencrypted.

Reviewer: FYI, looks like this is open on eu-latest. Doc wrong, or deployment is wrong?

Author: Bit of both... we opened it up during the private preview to help customers who could do unencrypted or two-way TLS, but not one-way TLS. Now that client certificate support is available that isn't necessary and we hope we won't need to  open 2889 on the higher zones. Perhaps we should clarify that unencrypted connections are an option for dedicated environments but not for our managed ones?

Reviewer: I think the doc is already clear on that (line below). eu-latest is just an anomaly, so I'm happy to leave this.

which it received the control packet which caused the violation.

In such instances MQTT clients must reconnect to be able to receive error messages from the error topic via the subscription. Error messages received after this reconnection
are from the previous session. This can lead to confusion when attempting corrective actions. Therefore, we highly recommend you to build a microservice which uses
the MQTT Service SDK to consume error messages, or use MQTT 5 for clients and make use of the reason codes feature.

#### Topic cleanup {#topic-cleanup}

The MQTT service will automatically remove topics which are no longer active. Topics are recognized as inactive when there are no subscriptions and
the internal publisher to the topic is closed. The publisher is responsible for publishing the modified MQTT service messages to the correct topic.
The publishers live within a cache, where the publisher expires after one hour. Due to this it can take up to an hour after removing all subscriptions from a topic
for it to be automatically deleted.

### Payload {#payload}

> The original MQTT messages are re-packed into MQTT Service message format which includes the original payload and additional metadata fields.

Reviewer: ```suggestion
The original MQTT messages map bidirectionally to the MQTT Service message format which includes the original payload and additional metadata fields.
```
because you can go both ways using the Java client

The **MQTT ClientID** field identifies the connected client. **ClientID** may consist of up to 128 alphanumeric characters.
Each client connecting to MQTT Service must have a unique client identifier, connecting a second client with the same identifier will result in the previous client's disconnection.

#### Quality of Service (QoS) {#quality-of-service-qos}

The {{< product-c8y-iot >}} implementation supports two levels of MQTT QoS:

* QoS 0: At most once:
    - The client just sends the message once (fire and forget).
    - No response from the server.
    - No guarantee that subscribers will receive the message.
* QoS 1: At least once:
    - The client awaits server acknowledgment for each published message.
    - The client should re-send the message if there was no acknowledgement from the server.
    - It is guaranteed that subscribers will receive a message that was acknowledged by the server.
    - Subscribers may receive more than one copy of a message.
* QoS 2: Exactly once:
    - not supported

For subscriptions, MQTT Service will deliver all messages in the QoS that the client defined when subscribing to the topic.

#### Clean session {#clean-session}

> MQTT Service requires clean session to be set to "1" (true). We cannot guarantee that disabling clean session will work reliably, hence we recommend you to always enable clean session.

Reviewer: ```suggestion
MQTT Service requires clean session to be set to "1" (true). Disabling clean session has no effect, meaning that a reconnecting client will not see any messages that it missed since the last disconnection.
```

For detailed instructions on adding and trusting a CA certificate, see [Managing trusted certificates](/device-management-application/managing-device-data/#managing-trusted-certificates).

#### Creating self-signed certificates

In order to self-sign the device certificates, the root CA certificate needs to be created.
Using the OpenSSL CLI tool, create a private key and then generate a self-signed root certificate from it.
```console
openssl genpkey -algorithm RSA -out ca.key
openssl req -x509 -new -nodes -key ca.key -sha256 -days 3650 -out ca.crt -subj "/C=UK/O=YourCompany/OU=YourOrg/CN=MQTTServiceCA"
```
Then create a private key for the device, generate the certificate signing request from this private key, and then sign the CSR.
```console
openssl genpkey -algorithm RSA -out client.key
openssl rsa -in client.key -out client-key.pem -outform PEM
openssl req -new -key client.key -out client.csr -subj "/C=UK/O=YourCompany/OU=YourOrg/CN=mqtt-client"
openssl x509 -req -in client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client.crt -days 3650 -sha256
cat client.crt ca.crt > client-chain.pem
```

> There are further instructions regarding creating self-signed CA, intermediate, and device certificates certificates under [Generating and signing certificates](/device-integration/device-certificates/#generating-and-signing-certificates).

Reviewer: This link is half-broken - path is fine, anchor goes nowhere. `/device-certificate-authentication/managing-trusted-certificates/` might be what you want?

Author: Fixed, along with a few others that didn't point to where they should have.

At this time, manual device registration is not supported in the MQTT Service. Devices must be registered through the auto-registration process. For more details on auto-registration, refer to [Auto-registration](/device-certificate-authentication/device-certificates#registering-devices-using-certificates) guide.
When connecting devices to the MQTT Service using certificates, the tenant ID must be included in the MQTT CONNECT packet in the user name field.
This is required to correctly identify the tenant.

#### Adding and trusting CA certificate

TLS trust anchors in the {{< product-c8y-iot >}} platform are defined per tenant.
To use device certificates for authentication, the CA or intermediate certificate that signs the device certificates must be uploaded to the platform and added to the tenant’s list of trusted certificates.
Additionally, the **Auto registration** field must be enabled when adding certificates.
For detailed instructions on adding and trusting a CA certificate, see [Managing trusted certificates](/device-certificate-authentication/managing-trusted-certificates).

> #### Creating self-signed certificates

Reviewer: So, it looks like this stuff is an alias to `/device-certificate-authentication/device-certificates/`  , although the latter is more verbose . Is the procedure for creating client certs for MQTT Service devices the exact same as the procedure for client certs for devices connecting over HTTP/Core MQTT/LWM2M ? If so, it ought to live there, with just a cross-reference.

And if we do that, there's a list at the top of "Device Certificates" that needs its enumeration updating; currently mentioning HTTP, (Core) MQTT and LWM2M.

Author: @rohitjoshi-c8y can explain why we duplicated some of the existing content. Agreed that the list in the other section should mention the MQTT Service.

Reviewer 2: Hi @rthrippleton, we decided to duplicate because the original doc is muddled with irrelevant details and does not read well. So, we decided to write it for the MQTT Service with clear instructions on creating and using client certificates.

Reviewer: I can't persuade you to instead fix the original doc? It's not a great customer experience to have us create two bits of doc about the same topic and purposes; ideally we want one canonical bit of doc on this topic, and what's been written in _this_ PR is suitable for that.

Author: I don't disagree that our doc is better :-) nor that we should only explain this stuff once. However, the risk is that if we start updating the old material we a) end up owning it and b) tread on toes and make ourselves unpopular. That's not a fight I want to have *right now* but maybe something we can look at for GA.

Reviewer: 107b936a5

Reviewer 3: At least we should create a ticket to have the original section updated accordingly. I completely agree with @rthrippleton that we should not duplicate text. If you like I can replace the original texts with yours and take the risk of ownership/annoying people :-).

Reviewer: I think it would require too much technical attention to make sure it's right - it's not a one-for-one replacement, the overlap is a lot more vague than that.

MQTT 3.1 and 3.1.1 clients:

* Clients only have access to the reason code describing the failure when using QoS 1 with acknowledgements and only for SUBSCRIBE packets, where the reason code is `0x80`.
* For PUBLISH packets, the client will be disconnected with no further information as per the MQTT specification.

In addition to the topic count, the MQTT Service also limits the size of the message backlog on each topic.
The message backlog contains all messages that have been published on the topic but not yet received and acknowledged by all subscribers to the topic.
When the backlog limit is reached, further attempts to publish to the topic will fail until some messages have been consumed.

Each message in a topic backlog also has a time-to-live (TTL) that starts at the moment the message is published.
When the TTL of a message expires, that messages will be deleted from the backlog regardless of whether all subscribers have received it or not.
MQTT clients do not receive any notification that messages have been discarded from a backlog due to TTL expiry.

> #### Error Topic {#error-topic}

Reviewer 3: ```suggestion
#### Error topic {#error-topic}
```

### MQTT 5.0 features {#mqtt-50-features}

Clients can connect using version 5.0 of the MQTT protocol.
Support for additional MQTT 5.0 features will be added in future releases.

### MQTT TLS certificates {#mqtt-tls-certificates}

#### Server certificates {#server-certificates}

The MQTT Service uses the same server certificates that are assigned to the main {{< product-c8y-iot >}} environment domain.
It always sends these certificates during TLS handshake to devices.
Moreover, {{< enterprise-tenant >}}s are not able to customize those certificates via the SSL Management feature.

#### Device (client) certificates {#device-certificates}

Using device certificates with the MQTT Service shares the same requirements as outlined in [Device certificates](/device-certificate-authentication/device-certificates#general-requirements-for-connecting-devices-with-certificates).
Additionally, auto-registration must be enabled when uploading the CA certificate to the platform.
At this time, manual device registration is not supported in the MQTT Service.
Devices must be registered through the auto-registration process.

> For more details on auto-registration, refer to [Auto-registration](/device-certificate-authentication/device-certificates#registering-devices-using-certificates) guide.

Reviewer: As a relative n00b, if I read that link in the context of the MQTT service, Registering devices using certificates / Auto registration    implies that connecting with MQTT for the first time creates a user, or device credentials, or a device .... ? (none of these happen in practice)

BTW, `/device-certificate-authentication/device-certificates/` is a bit stuck in the past - there's an enumeration at the top of the various ways you can SSL your way into the platform, and the only MQTT mentioned is the core MQTT.

> At this time, manual device registration is not supported in the MQTT Service.
> Devices must be registered through the auto-registration process.
> For more details on auto-registration, refer to [Auto-registration](/device-certificate-authentication/device-certificates#registering-devices-using-certificates) guide.

Reviewer: ```suggestion
```
Discussed with Rohit, we agreed that these sentences were more confusing than helpful, as they implied that devices would get created by the MQTT service.
